File: Common/Serialization/ArraySerializer.vb
Public Class ArraySerializer(Of T)
    Inherits Serializer(Of T())

    Public Sub New()
        MyBase.New(NameOf(Array))
    End Sub

    Public Overrides Sub SetT(Formatter As FormatterSetProxy, Obj As T())
        Formatter.Set(NameOf(Obj.Length), Obj.Length)
        For I As Integer = 0 To Obj.Length - 1
            Formatter.Set(Nothing, Obj(I))
        Next
    End Sub

    Public Overrides Function GetT(Formatter As FormatterGetProxy) As T()
        Dim Length As Integer
        Length = Formatter.Get(Of Integer)(NameOf(Length))

        Dim R = New T(Length - 1) {}
        For I As Integer = 0 To Length - 1
            R(I) = Formatter.Get(Of T)(Nothing)
        Next

        Return R
    End Function

End Class

File: Common/Serialization/ArraySerializer.cs
using System;

namespace Ks.Common
{
    public class ArraySerializer<T> : Serializer<T[]>
    {
        public ArraySerializer() : base(nameof(Array))
        {
        }

        public override void SetT(FormatterSetProxy Formatter, T[] Obj)
        {
            Formatter.Set(nameof(Obj.Length), Obj.Length);
            for (var I = 0; I < Obj.Length; I++)
            {
                Formatter.Set(null, Obj[I]);
            }
        }

        public override T[] GetT(FormatterGetProxy Formatter)
        {
            int Length = Formatter.Get<int>(nameof(Length));

            var R = new T[Length];
            for (var I = 0; I < Length; I++)
            {
                R[I] = Formatter.Get<T>(null);
            }

            return R;
        }
    }
}
